### DISCUSSION TITLE
Tasmota esp32 berry script to support Firebase push notification (FCM) function

### DISCUSSION THREAD

**Discussion by usrrsr** [arendst/Tasmota#21918] [Q&A]
is it possible in Tasmota esp32 berry script to send Firebase push notification (FCM) function so that same can be received on android app or so.

Regards,

**Comment by s-hadinger** [comment]
It depends on how simple or complex that API is. Sorry I have no experience with it. If it's just a simple https request, it should work

**Reply by sfromis** [reply]
As their SDK is not available for Berry, you should expect to have to implementing the protocol, instead of their API. Not sure how involved that would be, but maybe authentication is "more than just a https request".

If you can use Firebase via MQTT, that might be a good path to integration.

**Comment by usrrsr** [comment]
Thanks for response.

Well documented library exists here 

https://github.com/mobizt/FirebaseClient

but i dont know how to integrate , It will be win win situation if developers integrate it in build even custom building also acceptable

This will enable esp32  or esp8266  tasmotised module to send firebase cloud FCM messages to app capability to tasmota


Regards,

**Reply by s-hadinger** [reply]
When integrating in Tasmota, I always ask: what use case are you trying to achieve? Would this feature be useful only to users that also develop a mobile app? If so that's very close to none. 

Btw the lib you mention seems quite heavyweight. If a simple https request works, building a simple Betty library should be easy

**Comment by usrrsr** [comment]
Thanks for response
By the way i dont know about Betty library, how to make and use it

Each iot firmware like Tasmota, ESPHome, Rainmaker alone has very limited and not so user friendly and customizable GUI development environment. All such firmware should need supporting mobile app to enrich the HMI and user control interface to turn on off devices etc

In such cases when user controls any device by giving command may be over MQTT, In new android 34 sdk , you have to keep running your app running in background , else service running mqtt stuff will be destroyed and more so android services is being replaced with Receivers, 

This makes very difficult to send notification on mobile device when app is not running , Only way is firebase cloud messaging (FCM), almost all developers are shifting to that to enable there app connected to iot with mqtt or so to get notification even app is running


I think this features of firebase cloud messaging is to be integrated in Tasmota to take is further upper level than any other competitors and to enhance its utility further


Regards,

**Reply by s-hadinger** [reply]
I still don't see any use-case in your response. 

Who will develop that Android app you mention? You?

**Reply by barbudor** [reply]
I don't believe it is the rôle of Tasmota
Tasmota's rôle is to signal events to a MQTT broker and then let the backend handle that
That's what I already believe about email, telegram, influxdb...

**Reply by usrrsr** [reply]
> I still don't see any use-case in your response.
> 
> Who will develop that Android app you mention? You?

All developers like son-off developer, tuya , remote switch developer will develop such app to integrate tasmotised device with customized gui like Blynk, ESPrainmaker but they dont provide much flexibility to app developer
Well its all about notifying the user about the tasmota events since app no more have continuous connection with mqtt 

but leave it , i will find out the way out and post here the solution for others to start directly

**Reply by s-hadinger** [reply]
Please understand that we avoid integrating features if there is no explicit use-case. If you want to work on it fine with me.

**Reply by Jason2866** [reply]
For such use cases there are Home Automation softwares available. Open Source too. Tasmota and other firmwares which can replace the vendor firmware on the device plays nicely with Home Automation softwares. All actual Open Source Home Automation systems do support controlling from away. The GUI from this softwares are all very user friendly. Probably this softwares are able to communicate with firebase
Tasmota was and is designed to be used with a sort of Home Automation Hub. Having an APP directly communicating with the devices makes no sense at all in this concept.

Firebase concept does not fit to Tasmota. Tasmota is designed to use mqtt to communicate.
Furthermore Tasmota has the goal to get away from vendor lock in Cloud Solutions. Firebase is just that. To be very clear there will be no development done to support. A provided PR for supporting Firebase which needs changes in the "core" of Tasmota or increases code size when not active will have a low chance for a merge.

**Reply by sfromis** [reply]
What I'd be inclined to do in such a case would be to use Node-Red on my backend server, having a small flow to connect the dots for interfacing with Firebase. Or whatever platform works well with Firebase and MQTT.

**Comment by usrrsr** [comment]
I found the way around to get firebase fcm cloud messaging http v1 access token key berry code in tasmota/Berry_playground located here
https://github.com/tasmota/Berry_playground/blob/main/modules/googleapis/README.md

For FCM http v1 access token use following code use case example
google.json is service account json file from firebase console
```
load("googleoauth.be")
var auth = google_oauth("/google.json", "https://www.googleapis.com/auth/cloud-platform")
var cl = webclient()
cl.begin('https://www.googleapis.com/auth/firebase.messaging')
auth.add_access_key(cl)
cl.GET()
print(cl.get_string) # this will show an error, but not an auth error.
print(auth.access_token)
```


To send the notification using curl 

curl -X POST https://fcm.googleapis.com/v1/projects/pump-controller-c7f9b/messages:send -H "Authorization: Bearer [paste auth.access_token here]" -H "Content-Type: application/json" -d '{"message":{"topic":"246F28B49F88","notification": {"title":"title","body":"body"},"data":{"story_id":"story_12345"}}}'
Thanks for support

Regards,
